### PR TITLE
fix(router): tree bug where loop index is not decremented.

### PR DESCRIPTION
**Pull request by kristiansvalland**
Also add test to validate fix.

fixes https://github.com/gin-gonic/gin/issues/3459

### CODE REVIEW

File: tree.go [hunk 1]
```diff
@@ -462,6 +462,7 @@ walk: // Outer loop for walking the tree
 						for l := len(*skippedNodes); l > 0; {
```
Comment (thinkerou) [inline tree.go:462]: use `for l := len(*skippedNodes); l >0; l--)`?

maybe rename `l` to `length`, because `l` like `1`

File: tree.go [hunk 2]
```diff
@@ -579,6 +580,7 @@ walk: // Outer loop for walking the tree
 				for l := len(*skippedNodes); l > 0; {
 					skippedNode := (*skippedNodes)[l-1]
 					*skippedNodes = (*skippedNodes)[:l-1]
+					l--
```
Comment (thinkerou) [inline tree.go:583]: as above

File: tree.go [hunk 3]
```diff
@@ -636,6 +638,7 @@ walk: // Outer loop for walking the tree
 			for l := len(*skippedNodes); l > 0; {
 				skippedNode := (*skippedNodes)[l-1]
 				*skippedNodes = (*skippedNodes)[:l-1]
+				l--
```
Comment (thinkerou) [inline tree.go:641]: as above

File: tree.go (hunk 1 above)
Comment (kristiansvalland) [reply tree.go]: Much cleaner. Fixed :)

**Review by thinkerou — APPROVED** [review]
lgtm, thanks!

**Review by appleboy — APPROVED** [review]

File: tree.go (hunk 3 above)
Comment (beage666) [reply tree.go]: Up date

**Pull request merged**